### PR TITLE
RavenDB-18920 - Ability to download logs

### PR DESCRIPTION
**Pull request by grisha-kotler**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18920

### Additional description

Ability to download logs that are stored in the logs folder

### Type of change

- New feature

### How risky is the change?

- Low 

### Testing 

- It has been verified by manual testing

### UI work

- It requires further work in the Studio

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs [hunk 1]
```diff
@@ -77,5 +81,47 @@ public async Task RegisterForLogs()
...
+
+                                    await using (var entryStream = entry.Open())
+                                    {
+                                        await fs.CopyToAsync(entryStream);
+                                    }
+                                }
+                            }
+                            catch
+                            {
+                                // something went wrong with copying the file, will continue to the next one
```
Comment (ml054) [inline src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: should we create fake zip entry with errors in such case?

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs [hunk 2]
```diff
@@ -77,5 +81,47 @@ public async Task RegisterForLogs()
...
+
+        [RavenAction("/admin/logs/download", "GET", AuthorizationStatus.Operator)]
+        public async Task Download()
+        {
+            var contentDisposition = $"attachment; filename={DateTime.UtcNow:yyyy-MM-dd H:mm:ss} - Node {ServerStore.NodeTag} - Logs.zip";
+            HttpContext.Response.Headers["Content-Disposition"] = contentDisposition;
+
+            using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            {
+                await using (var ms = new MemoryStream())
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: Why are you writing this to memory and not to the request output stream?

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: yes

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: changed to temp file stream

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: done

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs [hunk 3]
```diff
@@ -77,5 +84,68 @@ public async Task RegisterForLogs()
...
+            HttpContext.Response.Headers["Content-Type"] = "application/zip";
+
+            var adminLogsFileName = $"admin.logs.download.{Guid.NewGuid():N}";
+            var adminLogsFilePath = ServerStore._env.Options.DataPager.Options.TempPath.Combine(adminLogsFileName);
+
+            var from = GetDateTimeQueryString("from", required: false);
+            var to = GetDateTimeQueryString("to", required: false);
+
+            using (var stream = new TempFileStream(SafeFileStream.Create(adminLogsFilePath.FullPath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.ReadWrite, 4096,
+                       FileOptions.DeleteOnClose | FileOptions.SequentialScan)))
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: DeleteOnClose does not work properly on Linux. Create 'DeleteOnCloseFileStream' maybe (we have DeleteOnCloseZipArchive already) and handle it properly.

Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: IIRC, they added support for that in the dotnet API

File: src/Raven.Studio/typescript/common/bindingHelpers/datePickerBindingHandler.ts [hunk 4]
```diff
@@ -44,6 +44,8 @@ class datePickerBindingHandler {
         const date : moment.Moment = ko.unwrap(valueAccessor());
         if (date) {
             $(element).data("DateTimePicker").date(date);
```
Comment (ml054) [inline src/Raven.Studio/typescript/common/bindingHelpers/datePickerBindingHandler.ts]: please simplify this code.

File: src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts [hunk 5]
```diff
@@ -448,6 +547,33 @@ class adminLogs extends viewModelBase {
...
+        this.downloadLogsValidationGroup.errors.showAllMessages(false);
+    }
+
+    onDownloadLogs() {
+        if (!this.isValid(this.downloadLogsValidationGroup)) {
+            return;
+        }
+
+        const url = endpoints.global.adminLogs.adminLogsDownload;
+        const argsPart = "?from=" + this.startDateToUse() + "&to=" + this.endDateToUse();
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts]: do we need that if we have [name=from], [name=to]?

Comment (ml054) [reply src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts]: + should we escape uri compoment?

File: src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts [hunk 6]
```diff
@@ -448,6 +547,33 @@ class adminLogs extends viewModelBase {
     onOpenSettings() {
         this.loadLogsConfig();
     }
+
+    onOpenDownload() {
+        this.startDate(null);
+        this.endDate(null);
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts:551]: do we need this?

File: src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts [hunk 7]
```diff
@@ -134,6 +157,28 @@ class adminLogs extends viewModelBase {
...
+            return this.useMinStartDate() ? null : this.startDate().utc().format(generalUtils.utcFullDateFormat);
+        });
+
+        this.endDateToUse = ko.pureComputed(() => {
+            return this.useMaxEndDate() ? null : this.endDate().utc().format(generalUtils.utcFullDateFormat);
+        });
+    }
+
+    dateFormattedAsUtc(localDate: moment.Moment) {
+        return ko.pureComputed(() => {
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts]: I think we don't need to return pure computed here.

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs [hunk 8]
```diff
@@ -77,5 +84,68 @@ public async Task RegisterForLogs()
...
+            HttpContext.Response.Headers["Content-Disposition"] = contentDisposition;
+            HttpContext.Response.Headers["Content-Type"] = "application/zip";
+
+            var adminLogsFileName = $"admin.logs.download.{Guid.NewGuid():N}";
+            var adminLogsFilePath = ServerStore._env.Options.DataPager.Options.TempPath.Combine(adminLogsFileName);
+
+            var from = GetDateTimeQueryString("from", required: false);
+            var to = GetDateTimeQueryString("to", required: false);
+
+            using (var stream = new TempFileStream(SafeFileStream.Create(adminLogsFilePath.FullPath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.ReadWrite, 4096,
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: why this is wrapped in TempFileStream?

File: src/Raven.Studio/typescript/common/bindingHelpers/datePickerBindingHandler.ts (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/common/bindingHelpers/datePickerBindingHandler.ts]: done
`$(element).data("DateTimePicker").date(date || null);`

File: src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts (hunk 6 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts:551]: yes - this clears the date picker content

File: src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts (hunk 7 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts]: done

File: src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts (hunk 5 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/adminLogs.ts]: right - we actually don't need the argsPart at all - removed

**Review by ml054 — APPROVED** [review]

File: src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs (hunk 8 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Handlers/Admin/AdminLogsHandler.cs]: done

**Pull request merged**